### PR TITLE
Support Custom Localization Mappings

### LINKED ISSUE #85: Support CUE for complex mappings in Localization
## Description 

Currently the mappings provided by localization are limited to the elements of a docker image reference.

In order to allow for arbitrary mapping strategies we should explore the use of CUE as a means by which to express the mapping.  

## Possible Approach

The below outlines one way in which the api for this might look. We provide the mapping details using a ConfigData file as usual but instead of specifying a particular resource to use for the localization, we pass the entire component-descriptor into the CUE script as a data variable. This can then be used within the `mapping` script to generate appropriate output. The output is returned to the controller and patched into the specified file as the given path.

```yaml
apiVersion: config.ocm.software/v1alpha1
kind: ConfigData
metadata:
  name: ocm-config
  labels:
    env: test
localization:
# the component descriptor is passed to the cue mapping expression
# the mapping expression generates some data which is returned to the localization controller via the "out" variable.
# the contents of out are inserted into the file "deploy.yaml" at the path "spec.servicePresets.images"
- file: bizx_environment.yaml
  path: spec.servicePresets.images
  mapping: |
    package main

    import (
      "encoding/yaml"
      "strings"
    )

    # the variable "component" is populated
    # with the component descriptor data
    images: {
      for r in component.resources {
        let name = strings.Replace(r.name, "bizx-", "", -1)
        "\(name)": r.access.imageReference
      }
    }

    out: yaml.Marshal(images)
```

For a component descriptor containing the following resources:

```yaml
resources:
  - name: bizx-logstash
    version: "0.1.5-7.17.6"
    type: ociImage
    access:
      type: ociRegistry
      imageReference: sf-docker-golden.common.cdn.repositories.cloud.sap/bizx-sidecars/logstash:0.1.5-7.17.6
  - name: bizx-grafana
    version: "0.1.5-7.17.6"
    type: ociImage
    access:
      type: ociRegistry
      imageReference: sf-docker-golden.common.cdn.repositories.cloud.sap/bizx-sidecars/grafana:0.1.5-7.17.6
```
The above cue mapping would generate the following output:

```yaml
logstash: sf-docker-golden.common.cdn.repositories.cloud.sap/bizx-sidecars/logstash:0.1.5-7.17.6
grafana: sf-docker-golden.common.cdn.repositories.cloud.sap/bizx-sidecars/grafana:0.1.5-7.17.6
```

Which would be patched into the file `bizx_environment.yaml`  at the path `spec.servicePresets.images`  resulting in the following localized resource:

```yaml
apiVersion: sapsf.com/v1alpha1
kind: BizxEnvironment
spec:
  servicePresets:
    images:
      logstash: sf-docker-golden.common.cdn.repositories.cloud.sap/bizx-sidecars/logstash:0.1.5-7.17.6
      grafana: sf-docker-golden.common.cdn.repositories.cloud.sap/bizx-sidecars/grafana:0.1.5-7.17.6
```

### LINKED ISSUE #86: Ensure all image reference parameters are exposed for Localization
## Description

For a given image reference composed of {registry}/{repository}/{image}:{tag}@{digest} the Localization controller should expose all parameters in the spec for Localization.

### PR DESCRIPTION
**Pull request by phoban01**
## Description

This pull request enables the execution of custom Localization mappings using the CUE configuration language.

The component descriptor object is accessible within the transform script via the `component` variable.

The value of the `out` variable will be injected into the file at `mapping.path`. The value in `out` should be JSON encoded.

```yaml
- file: custom_resource.yaml
  mapping:
    path: spec.values
    transform: |-
        import "encoding/json"
        values: [for x in component.resources {
          name: "\(x.name)-\(x.digest.hashAlgorithm)-\(x.version)"
          image: x.access.globalAccess.ref
        }]
        out: json.Marshal(values)
```
For a component descriptor containing a single resource, the above mapping will produce the following result:

```yaml
apiVersion: acme.org/v1alpha1
kind: MyResource
metadata:
  name: my-resource
  namespace: default
spec:
  values:
  - image: ghcr.io/mandelsoft/cnudie/component-descriptors/github.com/vasu1124/introspect
    name: introspect-image-sha256-1.0.0
```

One caveat to observe is that the target path *must* exist in order for the substitution to be performed.

Resolves #85 

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

### CODE REVIEW

File: controllers/mutation_reconcile_looper.go [hunk 1]
```diff
@@ -310,6 +312,17 @@ func (m *MutationReconcileLooper) getSnapshotBytes(ctx context.Context, snapshot
 func (m *MutationReconcileLooper) createSubstitutionRulesForLocalization(componentDescriptor v1alpha1.ComponentDescriptor, config configdata.ConfigData) (localize.Substitutions, error) {
 	var localizations localize.Substitutions
 	for _, l := range config.Localization {
+		if l.Mapping != nil {
+			res, err := m.compileMapping(componentDescriptor, l.Mapping.Transform)
```
Comment (Skarlso) [inline controllers/mutation_reconcile_looper.go:316]: This is pretty cool!

**Review by Skarlso — APPROVED** [review]

**Comment by Skarlso** [comment]
So now, this can utilize all the power of Cue, right?

**Comment by Skarlso** [comment]
@phoban01 It looks like your commit isn't signed. :/ I can't merge until then.

**Comment by phoban01** [comment]
Argh... signed now.

Yep you can use whatever CUE you like @Skarlso.

I'll update the README later today.

**Pull request merged**